delta-lake/src/main/java/org/apache/iceberg/delta/BaseMigrateDeltaLakeTableAction.java

```diff
@@ -0,0 +1,275 @@
+      Map<String, List<Action>> deltaLakeActionMap =
+          actions.stream()
+              .filter(action -> action instanceof AddFile || action instanceof RemoveFile)
+              .collect(Collectors.groupingBy(a -> a.getClass().getSimpleName()));
+
+      List<DataFile> filesToAdd = Lists.newArrayList();
+      List<DataFile> filesToRemove = Lists.newArrayList();
+      for (Action action : Iterables.concat(deltaLakeActionMap.values())) {
+        DataFile dataFile = buildDataFileFromAction(action, table, spec);
+        if (action instanceof AddFile) {
+          filesToAdd.add(dataFile);
+        } else if (action instanceof RemoveFile) {
+          filesToRemove.add(dataFile);
+        } else {
+          throw new ValidationException(
+              "The action %s's is unsupported", action.getClass().getSimpleName());
+        }
+      }
+
+      if (filesToAdd.size() > 0 && filesToRemove.size() > 0) {
+        // Overwrite_Files case
+        OverwriteFiles overwriteFiles = table.newOverwrite();
+        filesToAdd.forEach(overwriteFiles::addFile);
+        filesToRemove.forEach(overwriteFiles::deleteFile);
+        overwriteFiles.commit();
+      } else if (filesToAdd.size() > 0) {
+        // Append_Files case
+        AppendFiles appendFiles = table.newAppend();
+        filesToAdd.forEach(appendFiles::appendFile);
+        appendFiles.commit();
+      } else if (filesToRemove.size() > 0) {
+        // Delete_Files case
+        DeleteFiles deleteFiles = table.newDelete();
+        filesToRemove.forEach(deleteFiles::deleteFile);
+        deleteFiles.commit();
+      }
+    }
+  }
+
+  private DataFile buildDataFileFromAction(Action action, Table table, PartitionSpec spec) {
+    String path;
+    Optional<Long> nullableSize;
+    Map<String, String> partitionValues;
+    long size;
+
+    if (action instanceof AddFile) {
+      AddFile addFile = (AddFile) action;
+      path = addFile.getPath();
+      nullableSize = Optional.of(addFile.getSize());
+      partitionValues = addFile.getPartitionValues();
+    } else if (action instanceof RemoveFile) {
+      RemoveFile removeFile = (RemoveFile) action;
+      path = removeFile.getPath();
+      nullableSize = removeFile.getSize();
+      partitionValues = removeFile.getPartitionValues();
+    } else {
+      throw new IllegalStateException(
+          String.format(
+              "Unexpected action type for Delta Lake: %s", action.getClass().getSimpleName()));
+    }
+
+    if (partitionValues == null) {
+      // For unpartitioned table, the partitionValues should be an empty map rather than null
+      throw new ValidationException("File %s does not specify a partitionValues", path);
+    }
+
+    String fullFilePath = deltaLog.getPath().toString() + File.separator + path;
+    FileFormat format = determineFileFormatFromPath(fullFilePath);
+
+    Metrics metrics =
+        FileMetricsReader.getMetricsForFile(table, fullFilePath, format, this.hadoopConfiguration);
```

Comment:
given the fact that we can get both metrics and file size from input file, this refactoring to `FileMetricsReader` might not be needed

Reply:
Thank you for your suggestions. I refactored the code to get rid of the `FileMetricsReader` class. Currently, I relied on the FileIO returned by `table.io()` to generate the inputFile and use it to get metrics.